File: java/src/main/java/com/my/ippv/PhoneNo.java
package com.my.ippv;

public class PhoneNo {
    class Node {
	public String iso_;
        public Node[] child_;
        Node() {
	    iso_ = null;
            child_ = new Node[10];
            for (int i = 0; i < child_.length; ++i)
                child_[i] = null;
        }
    }

    Node root_ = new Node();

    public void add(String code, String iso) {
        var node = root_;
        for (int i = 0; i < code.length(); ++i) {
            int index = (int)code.charAt(i) - (int)'0';
            var tmp = node.child_[index];
            if (tmp == null) {
                tmp = new Node();
                node.child_[index] = tmp;
            }
	    if (i == code.length() - 1) {
	        tmp.iso_ = iso;
	    }
            node = tmp;
        }
    }

    public Result match(String phone) {
        var node = root_;
        var res = new Result();
        for (int i = 0; i < phone.length(); ++i) {
            int index = (int)phone.charAt(i) - (int)'0';
            node = node.child_[index];
            if (node == null)
                break;
	    if (node.iso_ != null) {
		    res.iso_ = node.iso_;
		    if (res.cc_ == 0)
			    res.cc_ = i + 1;
		    else if (res.ld_ == 0)
			    res.ld_ = i + 1;
		    else res.ac_ = i + 1;
	    }
        }

        return res;
    }
}

File: kotlin/src/main/kotlin/com/y/ippv/PhoneNo.kt
package com.y.ippv

class PhoneNr {
    data class Node(val iso: String?, val child: Array<Node?> = Array(10) { null })
    private val root_: Node = Node(null)

    fun add(cc: String, iso: String) {
        var node = root_
        cc.mapIndexedNotNull { i: Int, c: Char ->
            val index = c.code - '0'.code
            var tmp = node.child[index]
            if (tmp == null) {
                tmp = Node(if (i == cc.length - 1) iso else null)
                node.child[index] = tmp
            }
	    node = tmp
	    null
        }
    }

    fun match(phone: String): Result {
        var node: Node? = root_
	var len = 0
        var res = Result()
        for (ch in phone) {
            node = node!!.child[ch.code - '0'.code]
            if (node == null) {
		    break;
	    }
	    len++
	    if (node.iso != null) {
		    res.iso = node.iso
		    if (res.cc == 0)
		    	res.cc = len
		    else if (res.ld == 0)
		    	res.ld = len
		    else 
		    	res.ac = len
	    }
        }

        return res
    }
}

data class Result(var cc: Int = 0, var ld: Int = 0, var ac: Int = 0, var iso: String? = null)
